lib/loader.js

```diff
@@ -112,13 +154,6 @@ module.exports = function(content, map) {
 		if(sourceMap && result.map) {
 			// add a SourceMap
 			map = result.map;
-			if(map.sources) {
```

Comment:
Can we safely remove this ? `sourceRoot` should be empty (not sure if it can be non existent) and the `file` and `sources` paths are normalized here (L116+). Is this a bug ? :) What happens on windows platforms ?

Please elaborate why you removed it

Reply:
Short answer: **no, the removal of this code must remain.**

---

`sourceRoot` is something that the developer responsible for creating / maintaining the build system should be managing. We should not be modifying it.

Whatever is consuming the map will prepend the value of sourceRoot to each entry in sources to create a URL. If it is not an absolute URL, then it treats the path as a relative URL that is relative to the **map** not the CSS file, though this may be one and the same if the map is inlined into the CSS file.

I need to double check this next part, but I think I remember seeing that some uses of `sourceRoot` will modify the resulting paths in the `sources` array, meaning it has an impact on how future map generators and consumers handle including existing maps into newly generated maps.

---

The normalizing you are referring to attempted to do multiple things. The first is account for use of Webpack's currentRequest and remainingRequest in the `from` and `to` fields. That is going away and is no longer relevant.

The remaining normalizing attempts to treat `sources` as file system paths, which is wrong. Under normal use, they are URLs being consumed by a web browser to request source files from a web server.

The `path` module is one of node's core modules for working with file system paths. not URL paths.
---

In a completely closed build system, it may be possible to use file system paths, but Webpack and its loaders are not closed, they are open, as are the processors that the loaders and plugins consume. Since URLs is what the spec defines, we must provide URLs for other tools to work as expected.

source map writers (style-loader, mini-css-extract-plugin, etc.) may provide an option to output the final maps as absolute file system paths because they will be capable of resolving relative URLs to relative file system paths, and ultimately, absolute file system paths. in this scenario, the writer would want to remove the sourceRoot property or set it to be the `file:` URL protocol/URI scheme.

The are multiple reasons for this, but each have to do with the assumption that the file writer is the last thing to modify the resource as it will take into account the output destination. Remember, changing the location from the source directory to the build directory is also a transformation, so the writer should be creating an updated map of its own. We see this working correctly with [`gulp.dest()`]( https://github.com/gulpjs/gulp/blob/49b5aca613b33c5b626ae68c03a385f25c142f55/docs/API.md#gulpdestpath-options ) and [`gulpSourcemaps.write()`]( https://github.com/gulp-sourcemaps/gulp-sourcemaps/tree/492d937ffa0d5aa62a0d5b99ceb7deb2f472c7ad#write-options ).

---

For whatever reason, the Webpack community has tried to treat the URLs as absolute file system paths at every step. They also treat `sourceRoot` differently, when it shouldn't be treated at all. Some try to make paths relative to the Current Working Directory under the assumption that it is also the host/document root directory. Many of these loaders and plugins are passing mangled maps to tools like node-sass, postcss, and babel. They in turn, output bad maps or may ignore the mangled map entirely, and it continues to get worse with each node module touching the source map like this.

**Why has it kind of worked up until this point?**

-   sometimes the current working directory actually is the host root (but this is never guaranteed, and we shouldn't be using file system paths anyway)
-   when you embed source content in the map, the browser uses that instead of trying to fetch the source file. This why a number of guide may recommend this, because "it just works"
    -   Developers are left to make reasonably educated guesses on the actual location of the source file based on what ever value makes it into the sources array. If this value is an unmangled absolute file system path, this gives the appearance of it working correctly when it is not.
    -   Certain developers tools and plugins in the browser may be able to actually access the file system (by default or through permissions and configuration), but since Webpack is not a closed system, we can never assume this.

I am absolutely able to support this through experience, results, and referencing the specification and other places I've made inquiries on for this topic.

This is part of why I can say with absolute confidence that this is not just a single loader's problem. I'm starting with css-loader and sass-loader, but I see a lot of work to be done before this gets to where it needs to be in the webpack community.

Apologies for the rant! I hope it didn't come off as defensive. I'm just a bit nervous because I have seen past open issues effectively say "there is nothing wrong with this repository's code, go open an issue in the other module's repo". Since the problem spans multiple code bases, I will need support from contributors in each of those repositories. So part of this is trying to make the argument for it as strong and as clear as possible.

I will be happy to explain/converse more if I didn't clear it up. I'm unfortunately better at speaking code to computers than speaking English to people, so by all means, ask me as much as you need to! I'll do my best to be thorough and constructive.

Reply:
> sourceRoot is something that the developer responsible for creating / maintaining the build system should be managing. We should not be modifying it.

'`webpack`' [`SourceMapDevToolPlugin`](https://github.com/webpack/webpack/blob/master/lib/SourceMapDevToolPlugin.js) modifies it in the end, to my (limited) knowledge a `map`'s `sourceRoot` returned fro a loader should be either `''` or `undefined`

> The normalizing you are referring to attempted to do multiple things. The first is account for use of Webpack's currentRequest and remainingRequest in the from and to fields. That is going away and is no longer relevant.

👍 

> Under normal use, they are URLs being consumed by a web browser to request source files from a web server.

> The path module is one of node's core modules for working with file system paths. not URL paths.

> In a completely closed build system, it may be possible to use file system paths, but Webpack and its loaders are not closed, they are open, as are the processors that the loaders and plugins consume. Since URLs is what the spec defines, we must provide URLs for other tools to work as expected.

> source map writers (style-loader, mini-css-extract-plugin, etc.) may provide an option to output the final maps as absolute file system paths because they will be capable of resolving relative URLs to relative file system paths, and ultimately, absolute file system paths. in this scenario, the writer would want to remove the sourceRoot property or set it to be the file: URL protocol/URI scheme.

`webpack` uses it's own (devtools) protocol `webpack://` for source maps, which are URL's but `sources` ideally should be absolute paths for the `SourceMapDevToolPlugin`to work correctly. That's what I was at least told when asking about it the last time :)

For the `style-loader` it may be a different story since we can inline the map in the `<style></style>` tags as an annontation comment, kind of 'bypassing'  `webpack` internals

Reply:
> For whatever reason, the Webpack community has tried to treat the URLs as absolute file system paths at every step. They also treat sourceRoot differently, when it shouldn't be treated at all. Some try to make paths relative to the Current Working Directory under the assumption that it is also the host/document root directory. Many of these loaders and plugins are passing mangled maps to tools like node-sass, postcss, and babel. They in turn, output bad maps or may ignore the mangled map entirely, and it continues to get worse with each node module touching the source map like this.


Again likely because of `SourceMapDevToolPlugin` and because nobody _really_ knows how to handle it properly by default nor for anykind of variation xyz :)

> This is part of why I can say with absolute confidence that this is not just a single loader's problem. I'm starting with css-loader and sass-loader, but I see a lot of work to be done before this gets to where it needs to be in the webpack community.

True, it's an general problem within this area atm

Reply:
> 'webpack' (SourceMapDevToolPlugin) modifies it in the end, to my (limited in particular) knowledge a map's sourceRoot returned from a loader should be either '' or undefined

The `SourceMapDevToolPlugin` does modify it. I'll need to look at it more closely again. There is a lot going on so it will take some time for me to figure it out. This plugin is one of the "writers" I plan to make contributions to in the future (if needed), but for the last loader/plugin to be fixed, it must receive a valid Source Map from everything that has come before it.

---

> webpack uses it's own (devtools) protocol webpack:// for source maps, which are URL's, but sources ideally should be absolute paths for the SourceMapDevToolPluginto work correctly. That's what I was at least told when asking about it the last time :)
> 
> For the style-loader it may be a different story since we can inline the map in the <style></style> tags as an annotation comment (\n//# sourceMappingURL=[url]) directly, kind of 'bypassing' webpack internals
> 
> So from a plugin/loader authors perpective we need to provide the correct map values the SourceMapDevToolPlugin needs rather then the 'final' result

I am aware of the devtools option. My understanding so far is that it is a wrapper around `SourceMapDevToolPlugin` for less complex use cases.

I have tried to get more information about `webpack://` through StackOverflow per the instructions presented in when opening an issue in Webpacks repo. I was met largely with silence and downvotes, which ultimately lead to the auto deletion of the question.

From my own investigation, the scheme only appears so that it can be a default value for two configuration options: 
https://github.com/webpack/webpack/blob/4f4a2bae33fa1f8b1954aa0bbfd5b9e3bef95cf8/lib/SourceMapDevToolPlugin.js#L70-L74

**`webpack:` is not a valid registered URI scheme ( <https://www.iana.org/assignments/uri-schemes/uri-schemes.xml> ).** No web browser would know what to do with it. Beyond this, there is no publically information available to me about why this is used by webpack's core code.

`SourceMapDevToolPlugin` requiring absolute file system paths as input is wrong, though I haven't seen any documentation stating this as a requirement. Even if `SourceMapDevToolPlugin` knows what to do with them, other modules, which loaders and plugins often consume, such as PostCSS, Sass, Babel, and NYC/Istanbul, are not expecting absolute paths, nor is there any requirement from the Source Maps specification for them to do so. This can contribute to further mangling of the source maps.

`SourceMapDevToolPlugin` outputting absolute file system paths is not something I would object to, so long as it is a configurable option that is disabled by default. This kind of behavior should be opt in. Since it is effectively the last piece of software to perform a transformation, it is ok for the output to be whatever we want it to be. In doing so, the developer using our plugin or loader takes responsibility for it being different from the specification.

The `style-loader` has unique considerations which I will need a lot more time to learn about. This is thanks in part to it serving blobs. Support for hot module replacement will also add to the complexity. `webpack-dev-server` (and the middleware it consumes) will probably need to be tackled at the same time as the `style-loader`. Since URLs are normally relative URLs to the file containing the map (even inlined maps), serving the blob is effectively another transformation since you are changing the file's depth/path relative to the host root directory.

---

> Again likely because of SourceMapDevToolPlugin and because nobody really knows how to handle it properly by default nor for anykind of variation xyz :)

I have seen tools in the Gulp community handle it correctly, though there are some that don't. There is no reason why Webpack can't ultimately be the same. I think I have a reasonably good idea on how to handle it, my efforts here are an attempt to start down that path.

---

I will be happy to consider any of my current knowledge wrong if I can be pointed to specifications or documentation. What I have seen in other issues in and out of this repo boiled down to "I think that's how it works" or "this person told me this was how it works, but didn't cite any sources." Unless that person wrote the spec, that isn't enough for me to back down. So far, I haven't found any sources proving me wrong, just opinions and guesses. So I had to get the ball rolling with a pull request of my own.

Reply:
@michael-ciniawsky - I just wanted to check in since it has been about a week. No rush in replying! I just wanted to make sure I had touched on everything you were looking for a response on. 

For reference, @evilebottnawi mentioned in an issue in the sass-loader repo that source maps are not expected to be tackled until the next major revision of the loader. Is that also true for the css-loader?

Should I [resolve conversation](https://help.github.com/articles/commenting-on-a-pull-request/#resolving-conversations) to collapse this series of comments and replies and return to the main thread of the PR for discussion?